### PR TITLE
[fix][ttl] ttl-insert-cb-sync: Allow tensor SSA uses past the next-acquire boundary (#536)

### LINKED ISSUE #536: Auto-inserted cb_pop_front hoisted past subsequent cb_wait_front in same DM thread
## Summary

When a single datamovement thread does N consecutive `cb.wait()` calls on the same DFB, all auto-inserted `cb_pop_front` calls are hoisted to AFTER the last use. Without intervening pops, `get_read_ptr()` returns the same slot 0 pointer for every wait, so every consumer use sources from slot 0; later slots are correctly produced by compute but never read.

Single core, single compute + single dm_write thread, no pipes, no multi-thread CB sharing. Deterministic.

## Reproducer

```python
from __future__ import annotations
import torch
import ttnn
import ttl

TILE = 32


def make_kernel():
    @ttl.operation(grid=(1, 1))
    def repro(out):
        shape = (1, 1)
        shared_cb = ttl.make_dataflow_buffer_like(out, shape=shape, block_count=2)
        out_cb = ttl.make_dataflow_buffer_like(out, shape=shape, block_count=2)

        @ttl.compute()
        def compute():
            with shared_cb.reserve() as v:
                v.store(ttl.math.fill(v, 7.0))
            with shared_cb.wait() as src, out_cb.reserve() as dst:
                dst.store(src)

            with shared_cb.reserve() as v:
                v.store(ttl.math.fill(v, 8.0))
            with shared_cb.wait() as src, out_cb.reserve() as dst:
                dst.store(src)

        @ttl.datamovement()
        def dm_read():
            pass

        @ttl.datamovement()
        def dm_write():
            blk = out_cb.wait()
            ttl.copy(blk, out[0:1, 0:1]).wait()
            # blk.pop()  # workaround: explicit pop here makes it pass
            blk = out_cb.wait()
            ttl.copy(blk, out[1:2, 0:1]).wait()
            # blk.pop()

    return repro


def _to_dev(t, device):
    return ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                           device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)


if __name__ == "__main__":
    default_size = ttnn.device.get_max_worker_l1_unreserved_size()
    device = ttnn.open_device(device_id=0, worker_l1_size=default_size - 90112)
    try:
        k = make_kernel()
        H = 2 * TILE
        out_t = _to_dev(torch.full((H, TILE), -42.0, dtype=torch.bfloat16), device)
        k(out_t)
        ttnn.synchronize_device(device)
        out_h = ttnn.to_torch(out_t)
        vals = [out_h[0, 0].item(), out_h[TILE, 0].item()]
        expected = [7.0, 8.0]
        print(f"out={vals}  expected={expected}  "
              f"{'PASS' if vals == expected else 'FAIL (auto cb_pop_front hoisted)'}")
    finally:
        ttnn.close_device(device)
```

Output:
```
out=[7.0, 7.0]  expected=[7.0, 8.0]  FAIL (auto cb_pop_front hoisted)
```

## Generated metal kernel (dm_write)

```cpp
cb_wait_front(out_cb, 1);
ptr_a = get_read_ptr(out_cb);            // slot 0 (L1 base)
noc_async_write_tile(0, accessor, ptr_a); // DRAM[0] <- slot 0  (= 7.0)
noc_async_write_barrier();

cb_wait_front(out_cb, 1);                // no pop yet — front still slot 0
ptr_b = get_read_ptr(out_cb);            // SAME pointer as ptr_a
noc_async_write_tile(1, accessor, ptr_b); // DRAM[1] <- slot 0  (= 7.0)
noc_async_write_barrier();

cb_pop_front(out_cb, 1);  // both pops emitted at the very end
cb_pop_front(out_cb, 1);
```

`compute()`'s codegen is fine — slot 0 receives 7.0 and slot 1 receives 8.0, both visible after the pushes. The bug is purely in the consumer-side scheduling: with no `cb_pop_front` between the two `cb_wait_front`/`get_read_ptr` pairs, the read pointer never advances past slot 0.

## Workaround

Adding explicit `blk.pop()` after each use in `dm_write` makes the reproducer pass — it forces the pop into the IR between the two waits, so the read pointer advances correctly.

## Suggested fix

Auto-insert `cb_pop_front` immediately after the last use of each `cb.wait()` block (or before the next `cb.wait()` on the same DFB), rather than coalescing all pops at the end of the function.

### LINKED ISSUE #556: [ttl] Coalesce consecutive cb_wait acquires into multi-tile cb_wait_front(N) with per-acquire src_idx
## Problem

A sequence of `N` consecutive `cb_wait` ops on the same DFB lowers to `N` separate `cb_wait_front(1)` / `cb_pop_front(1)` calls. The buffer exposes a single FIFO front pointer, so the auto-pop pass must place `cb_pop` ops in the same op order it observes consumer uses; out-of-declaration-order consumes (consumer reads tile 2 before tile 1) cannot be expressed without violating FIFO monotonicity of the front pointer.

## Reproducer

`test/python/test_auto_pop_push.py::test_reordered_consumes_violate_fifo_xfail` (branch `bnorris/auto-push-pop-fixes`).

`@pytest.mark.xfail(strict=True)`. Flips to PASS once this lands.

## Proposed change

Detect strictly-consecutive acquires on the same DFB sync class (same block, only `attach_cb` / arith constants between them) and lower the group as one `cb_wait_front(N)` plus per-acquire `src_idx` in `[0, N)` plus one `cb_pop_front(N)`. Each acquire then addresses its tile by index, decoupling consume order from release order.

TableGen extension: optional `num_tiles` and `tile_offset` attributes on `CBWaitOp` / `CBPopOp` (and their producer-side mirrors `CBReserveOp` / `CBPushOp`). `num_tiles=0` marks a subsumed acquire whose `cb_wait_front` is elided; the SSA result still anchors downstream `src_idx` indexing.

Lowering: `ConvertTTLToTTKernel.cpp` reads the new attributes; `ConvertTTLTileOpsToTTKernel.cpp` folds `tile_offset` into `flatSrcIndex`. Independent of #555, but the unified ownership simplifies the coalescing analysis.

Related: #536.

### PR DESCRIPTION
**Pull request by brnorris03**
Problem:
`ttl-insert-cb-sync` inserts `cb_pop` / `cb_push` after the last use of each acquired slot, but only searched up to the next same-class acquire on the same DFB. Tensor SSA uses can legitimately appear after a later acquire (a tile from `cb_wait t1` may be consumed after `cb_wait t2`). Those uses were missed, the pop was inserted right after the acquire, and the read pointer advanced before the data was read. The case_a / case_b reproducers from the issue 536 follow-up [comment](https://github.com/tenstorrent/tt-lang/issues/536#issuecomment-4372107754) miscompiled silently.

Independently, deferred-consume patterns (`t1 = cb.wait(); t2 = cb.wait(); use(t1); use(t2)`) lowered to N independent `cb_wait_front(k)` / `cb_pop_front(k)` calls. Because metal's wait/pop primitives are non-cumulative, the first pop advances the front before the producer has pushed enough tiles to satisfy the next read.

Fixes (~530 LOC):
1. Split `findLastOwnedUse`: direct CB uses keep the boundary, tensor SSA uses ignore it. Extend `findOwnedReleases`' upper bound to the last owned use.
2. New pass `ttl-coalesce-dfb-acquires` rewrites a maximal run of same-DFB acquires + their matched releases into the canonical tt-metal cumulative-wait shape (`cb_wait_front(N*k)` + per-block `tensor.extract_slice` views + `cb_pop_front(N*k)`). `addSliceOffset` already folds the slice offsets into the per-tile `src_idx` / `dst_idx`, so no lowering changes. Same template handles producer-side `cb_reserve`/`cb_push`. The detection rule is locally-checkable: an op between members terminates the group iff it touches our DFB or consumes a group member's result, or carries a region (the conservative envelope of "could trigger a release on the DFB"). This means matmul-style patterns where acquires on two DFBs are interleaved (`a1, b1, a2, b2`) coalesce per-DFB independently, since an acquire on a different DFB is benign. See `docs/development/DFBManagement.md` for the full correctness argument. Fixes #556.

Tests:
- test/python/test_auto_pop_push.py: device coverage including the case_a / case_b reproducers, an interleaved-baseline regression guard, and axis-coverage cases (fanout, mixed immediate/deferred, long chains, multi-DFB, multi-tile geometry, tight block_count, single-slot, long DM loops, producer/consumer in scf.for, multiple direct DFB uses on one acquire). The reordered-consume case (formerly xfail #556) now runs as a passing test.
- test/ttlang/Dialect/TTL/Transforms/insert_cb_sync.mlir: IR-level coverage of the deferred-use pattern for compute-thread consumers (flat and inside scf.for) and producers, plus a regression guard for the scf.for iter_args pattern.
- test/ttlang/Dialect/TTL/Transforms/coalesce_dfb_acquires.mlir: positive (consumer / producer / loop-body / matmul-style two-DFB interleaved), negative (interleaved consume, single-acquire-per-DFB alternation, pre-existing num_tiles), and idempotency coverage for the new pass.
- docs/development/DFBManagement.md: extended "DFB Sync Insertion" with the ownership criteria, release-placement invariants, and idempotency.

Future work:
- #555: encode DFB ownership in SSA on `ttl.copy` operands. Would collapse `findLastOwnedUse` to a single SSA walk and lift the DM-thread xfails. Requires (mechanical) updates to a large number of lit tests.

Fixes #536 follow-up. Fixes #556.

### CODE REVIEW

**Review by zoecarver — APPROVED** [review]
LGTM!

**Review by arichinsTT — APPROVED** [review]
Looks good!
